subprojects/model-core/src/test/groovy/org/gradle/model/internal/manage/projection/ModelSetModelProjectionTest.groovy

```diff
@@ -73,6 +76,20 @@ class ModelSetModelProjectionTest extends ProjectRegistrySpec {
         set.toArray(new NamedThing[2]).collect { it.name } == ['1', '2']
     }
 
+    def "can define and query subclass elements"() {
+        when:
+        mutate {
+            create(NamedSubThing) { name = '1' }
+            create(NamedSubThing) { name = '2' }
+        }
+
+        then:
+        def set = registry.realize(collectionPath, collectionType)
+        set*.name == ['1', '2']
+        set.toArray().collect { it.name } == ['1', '2']
+        set.toArray(new NamedSubThing[2]).collect { it.name } == ['1', '2']
+    }
+
```

Comment:
Could you please add some coverage for when elements of different types are mixed in the same `ModelSet`?